src/module.indexer/model/supply.stat.ts

```diff
@@ -0,0 +1,210 @@
+import { BurnHistory } from '@defichain/jellyfish-api-core/dist/category/account'
+import { JsonRpcClient } from '@defichain/jellyfish-api-jsonrpc'
+import { NetworkName } from '@defichain/jellyfish-network'
+import { Inject, Injectable } from '@nestjs/common'
+import { Indexer, RawBlock } from '@src/module.indexer/model/_abstract'
+import { MasternodeStatsMapper } from '@src/module.model/masternode.stats'
+import { SupplyStatMapper } from '@src/module.model/supply.stat'
+import { SupplyStatAggregation, SupplyStatAggregationMapper } from '@src/module.model/supply.stat.aggregation'
+import { HexEncoder } from '@src/module.model/_hex.encoder'
+import BigNumber from 'bignumber.js'
+
+type EmissionRateBound = 'GENESIS' | 'POST_GENESIS' | 'EUNOS'
+
+@Injectable()
+export class EmissionCalculator {
+  constructor (
+    private readonly network: NetworkName
+  ) { }
+
+  readonly EMISSION_RATES: { [key in EmissionRateBound]: (height: number) => number } = {
+    GENESIS: _ => 591_000_030,
+    POST_GENESIS: _ => 200, // AMK, BayFront, ClarkQuay, Dahkota: total DFI emission, doesnt matter emitting utxos or token
+    EUNOS: (currentHeight: number) => {
+      const eunosHeight = EmissionConsensusParams[this.network].EUNOS_HEIGHT
+      if (currentHeight < eunosHeight) {
+        throw new Error('Invalid emission rate calculation input')
+      }
+
+      if (currentHeight % 1 !== 0 || eunosHeight % 1 !== 0) {
+        throw new Error('Invalid emission rate calculation input, block height must be whole number')
+      }
+
+      const params = EmissionConsensusParams[this.network]
+      const reductionCount = Math.floor((currentHeight - eunosHeight) / params.EMISSION_REDUCTION_PERIOD)
+
+      let subsidySatoshi = params.EUNOS_BASE_BLOCK_SUBSIDY_SATOSHI
+
+      for (let i = 0; i < reductionCount; i++) {
+        // no float/decimal operation
+        const reduceAmount = subsidySatoshi * params.EMISSION_REDUCTION / BigInt(100_000)
+        subsidySatoshi -= reduceAmount
+      }
+
+      return Number(subsidySatoshi) / 100_000_000
+    }
+  }
+
+  calculate (height: number): number {
+    const fork = this.getEmissionRateBound(height)
+    return this.EMISSION_RATES[fork](height)
+  }
+
+  getEmissionRateBound (height: number): EmissionRateBound {
+    if (height === 0) {
+      return 'GENESIS'
+    }
+
+    const params = EmissionConsensusParams[this.network]
+    if (height >= params.EUNOS_HEIGHT) {
+      return 'EUNOS'
+    }
+
+    return 'POST_GENESIS'
+  }
+}
+
+interface EmissionConsensusParamsI {
+  PRE_EUNOS_BASE_BLOCK_SUBSIDY: number
+  EUNOS_HEIGHT: number
+  EUNOS_BASE_BLOCK_SUBSIDY_SATOSHI: bigint
+  EMISSION_REDUCTION: bigint
+  EMISSION_REDUCTION_PERIOD: number
+}
+
+const EmissionConsensusParams: { [key in NetworkName]: EmissionConsensusParamsI } = {
+  mainnet: {
+    EUNOS_HEIGHT: 894000,
+    PRE_EUNOS_BASE_BLOCK_SUBSIDY: 200,
```

Comment:
i think should standardize the base block subsidy unit as satoshi since the new one using satoshi
```suggestion
    PRE_EUNOS_BASE_BLOCK_SUBSIDY: BigInt(20000000000),
```

Reply:
Will completely drop this, not used, no calculation needed for pre Eunos emission anyway.